File: apps/desktop/src/lib/daemon.ts
// Thin typed wrappers around the Tauri command bridge (src-tauri/src/commands.rs). This is the
// ONLY place the frontend talks to the daemon: no `fs`, no `path`, no socket — every read/write
// crosses through these `invoke` calls (design §7).
// Ref: https://v2.tauri.app/develop/calling-rust/ and https://v2.tauri.app/develop/calling-frontend/
import { invoke, listen, type UnlistenFn } from "./tauriShim";
import type { BuildInfo } from "./versionInfo";
import { retryWhileLoading } from "./loadingRetry";

/** Mirrors `desktop_lib::status::DaemonStatus` (serde `rename_all = "snake_case"`). */
export type DaemonStatus = "connected" | "connecting" | "spawning" | "dead";

/** Mirrors `desktop_lib::dto::FileInfoDto`. */
export interface FileInfo {
	path: string;
	hash: string;
	kind: string;
	done: number;
	total: number;
}

/** Every synced document with its current projection hash (`ListFiles`). */
export function listFiles(): Promise<FileInfo[]> {
	return retryWhileLoading(() => invoke<FileInfo[]>("list_files"));
}

/** Current connectivity state to `txtodod`, queryable without waiting for an event. */
export function daemonStatus(): Promise<DaemonStatus> {
	return invoke("daemon_status");
}

/** Retries the connect/spawn sequence; wired to the reconnect banner's retry button. */
export function retryConnect(): Promise<DaemonStatus> {
	return invoke("retry_connect");
}

/** Subscribes to `daemon-status` events, pushed on every status change. */
export function onDaemonStatus(cb: (status: DaemonStatus) => void): Promise<UnlistenFn> {
	return listen<DaemonStatus>("daemon-status", (event) => cb(event.payload));
}

/** Mirrors `desktop_lib::dto::FileContentsDto` (`GetFile`'s response). */
export interface FileContents {
	path: string;
	text: string;
	hash: string;
	/** One entry per line of `text`: the line's task id, `""` for a blank line. Absent or empty
	 * from a daemon older than the field (task sidecar-task-ids) — use `taskIdAt`, not this. */
	task_ids?: string[];
}

/** The exact bytes the daemon holds for one workspace-relative document path. */
export function getFile(path: string): Promise<FileContents> {
	return retryWhileLoading(() => invoke<FileContents>("get_file", { path }));
}

/** A line addressed by line number and/or ULID task id (`""` when the line has none yet). Mirrors `desktop_lib::dto::TaskRefDto`. */
export interface TaskRef {
	line_number: number;
	task_id: string;
}

/**
 * One intent-level mutation `apply()` can make. Mirrors `desktop_lib::dto::MutationDto`
 * (`#[serde(tag = "kind", rename_all = "snake_case")]` — field names below are otherwise
 * untouched by serde, so they stay exactly as declared on the Rust side).
 */
export type Mutation =
	| { kind: "add"; line: string }
	| { kind: "complete"; task: TaskRef; today: string }
	| { kind: "edit"; task: TaskRef; new_line: string }
	| { kind: "move"; task: TaskRef; to_path: string }
	| { kind: "delete"; task: TaskRef; leave_blank: boolean }
	// The whole document, compare-and-swap against `base_hash` (`FileContents.hash`). Must be the
	// only mutation of its `applyMutations` call. See `$lib/todotxt/saveBuffer.ts`.
	| { kind: "replace"; base_hash: string; contents: string };

/** The one `MutationDto` variant the edit popover needs, as its own type so callers don't have to
 * narrow `Mutation`'s union. Structurally identical to `Mutation`'s `"edit"` arm. */
export type EditMutation = Extract<Mutation, { kind: "edit" }>;

/** Mirrors `desktop_lib::dto::ResolutionDto` (serde `rename_all = "snake_case"` on a fieldless
 * enum serializes as a bare string). `"merged"` keeps whatever is already in the file and only
 * clears the flag — see `ConflictReviewSheet`'s `resolve()` doc comment for why that is not the
 * same thing as this app's client-side merge *preview*. */
export type ResolveChoice = "mine" | "theirs" | "merged";

/** Result of `apply()`/`resolve()`. Mirrors `desktop_lib::dto::ApplyResultDto`. */
export interface ApplyResult {
	applied: number;
	hash: string;
	hlc_wall_ms: number;
	hlc_counter: number;
}

/** This app's build and, when it answers, the daemon's (`commands_version.rs::build_info`). */
export function buildInfo(): Promise<BuildInfo> {
	return invoke("build_info");
}

/** Intent-level mutations on one workspace-relative document; the daemon turns them into ops.
 * `workspaceRoot` is the root the caller's buffer was read under: the bridge refuses the call
 * (`workspace-changed:`) when the app has switched workspace since, so a buffer saved during a
 * switch can never land in the other workspace's file. Omit it for a call that has no buffer.
 * Tauri maps this camelCase key to the command's `workspace_root` argument:
 * https://v2.tauri.app/develop/calling-rust/#passing-arguments */
export function applyMutations(path: string, mutations: Mutation[], workspaceRoot?: string): Promise<ApplyResult> {
	return invoke("apply", { path, mutations, workspaceRoot: workspaceRoot ?? null });
}

/** Whole-line replacement on one workspace-relative document (`Apply`); the daemon derives
 * field-level ops and rejects a stale write against a moved/deleted line via `task`. Thin
 * convenience over `applyMutations` for the edit popover's single-mutation case. */
export function applyEdit(path: string, mutation: EditMutation): Promise<ApplyResult> {
	return applyMutations(path, [mutation]);
}

/** One recorded op, as surfaced on a `Change`. Mirrors `desktop_lib::dto::OpSummaryDto`. `op_id`
 * is a ULID, whose first 10 chars embed the op's millisecond timestamp
 * (https://github.com/ulid/spec) — there's no separate timestamp field on the DTO today. */
export interface OpSummary {
	seq: number;
	op_id: string;
	device: string;
	principal: string;
	kind: string;
	task_id: string;
	summary: string;
}

/** Mirrors `desktop_lib::dto::HistoryDto`. */
export interface HistoryResult {
	ops: OpSummary[];
}

/** Ops newest first for one path/task (`History`); the edit popover's footer uses `limit: 1`.
 * Tauri maps camelCase JS invoke args to the command's snake_case parameter names, so `taskId`
 * here reaches `commands::history`'s `task_id`. Ref: https://v2.tauri.app/develop/calling-rust/ */
export function history(path: string, taskId: string, limit: number): Promise<HistoryResult> {
	return invoke("history", { path, taskId, limit });
}

/** One `needs_review` flag raised by a change. Mirrors `desktop_lib::dto::ReviewFlagDto`. */
export interface ReviewFlag {
	task_id: string;
	line_number: number;
	mine: string;
	theirs: string;
}

/** One `Watch` event. Mirrors `desktop_lib::dto::ChangeDto`. */
export interface Change {
	path: string;
	hash: string;
	ops: OpSummary[];
	review: ReviewFlag[];
}

/** Ensures the one shared `Watch` stream (every document) is forwarding each change as a
 * `daemon-change` event (see `onDaemonChange`) — every listener filters to the path(s) it cares
 * about itself. Resolves once the stream is established, not when it ends — safe, and cheap, to
 * call more than once (e.g. once per open document): the backend only opens the stream on the
 * first call after a connection (tasks/desktop-concurrent-edit-loss root cause 3 — this used to
 * open one never-cancelled forwarder per call, so one change fired several `refreshDoc`s).
 * Ref: https://v2.tauri.app/develop/calling-rust/ */
export function watch(): Promise<void> {
	return invoke("watch");
}

/** Subscribes to `daemon-change` events forwarded from an active `watch()` stream. */
export function onDaemonChange(cb: (change: Change) => void): Promise<UnlistenFn> {
	return listen<Change>("daemon-change", (event) => cb(event.payload));
}

/** Open `needs_review` flags for one workspace-relative document (plan M4). The daemon is
 * authoritative for this list — callers should reconcile it against any client-accumulated state
 * on mount/refresh rather than trusting the accumulation alone (design §4.7). */
export function listConflicts(path: string): Promise<ReviewFlag[]> {
	return invoke("list_conflicts", { path });
}

/** Resolves one `needs_review` flag; maps to the daemon's `ResolveConflict` RPC, the same
 * `Apply(Edit)` path the edit popover uses, so attribution/sync/history stay automatic and the
 * op itself clears the flag server-side. */
export function resolveConflict(
	path: string,
	task: TaskRef,
	resolution: ResolveChoice
): Promise<ApplyResult> {
	return invoke("resolve", { path, task, resolution });
}

/** `notes.md` for one task's `ref:` directory (plan M5). Mirrors `desktop_lib::dto::NotesDocDto`.
 * `text`/`path` are empty until the first `editNotes` call — the daemon lazily mints the `ref:`
 * tag and directory on that first write (`crates/txtodo-daemon/src/refdir_ops.rs::ensure_ref_dir`),
 * never the client (design §7). */
export interface NotesDoc {
	path: string;
	text: string;
	hash: string;
}

/** `notes.md` for one task's `ref:` directory; the daemon resolves the task to its directory by
 * `task.task_id` alone — no slug is ever computed or sent by this client. */
export function getNotes(task: TaskRef): Promise<NotesDoc> {
	return invoke("get_notes", { task });
}

/** Whole-document replacement of one task's `notes.md`; lazily creates the `ref:` tag and
 * directory on the first call for a task that doesn't have one yet (plan §3.2.4). */
export function editNotes(task: TaskRef, newText: string): Promise<ApplyResult> {
	return invoke("edit_notes", { task, newText });
}

/** Absolute workspace root, for the detail view's footer (display only — see
 * `src-tauri/src/commands.rs::workspace_root`'s doc comment for why this is still design-§7-safe).
 * Reflects whichever workspace `switchWorkspace` last selected, not a fixed startup value. */
export function workspaceRoot(): Promise<string> {
	return invoke("workspace_root");
}

/** Advisory-only: true when no `txtodo skill install` has run yet on this machine. Mirrors
 * `txtodo-cli`'s `doctor` skill row and `txtodo-tui`'s status-line hint. */
export function skillHintNeeded(): Promise<boolean> {
	return invoke("skill_hint");
}

/** One entry in the device-global workspace registry (ADR 0025). Mirrors
 * `desktop_lib::dto::WorkspaceInfoDto`. */
export interface WorkspaceInfo {
	id: string;
	root: string;
	added_at_ms: number;
	root_exists: boolean;
	has_state: boolean;
	/** How far the daemon is through opening it (task daemon-early-bind); "unknown" from an older daemon. */
	load_state: "queued" | "loading" | "ready" | "failed" | "unknown";
	/** Why the last open failed; empty unless `load_state` is "failed". */
	load_error: string;
	/** The user's default workspace (task default-workspace): labelled, and never removable. */
	is_default?: boolean;
}

/** Every registered workspace, oldest first. */
export function listWorkspaces(): Promise<WorkspaceInfo[]> {
	return invoke("list_workspaces");
}

/** Registers `root` (idempotent) without switching to it or opening it. */
export function addWorkspace(root: string): Promise<WorkspaceInfo> {
	return invoke("add_workspace", { root });
}

/** Un-registers a workspace id; never touches its `.txtodo/` state on disk. The daemon refuses
 * this for whichever workspace `root` names as the currently switched-to one — switch away
 * first. `false` means `id` was never a known workspace; `true` covers both a fresh removal and
 * removing an already-removed id (`Registry::remove` is an idempotent upsert-tombstone). */
export function removeWorkspace(id: string, root: string): Promise<boolean> {
	return invoke("remove_workspace", { id, root });
}

/** Points every future daemon call at `root` instead — no reconnect. Callers must also re-fetch
 * (and re-`watch`) anything already loaded against the old workspace: this only changes the
 * selector attached to calls made *after* it resolves, not any stream already open. */
export function switchWorkspace(root: string): Promise<void> {
	return invoke("switch_workspace", { root });
}

/** One `op_log_all()` row: an `OpEvent` (see `src/devices/types.ts`) tagged with its source
 * workspace — the field `OpLogEntry` itself doesn't carry (task
 * `desktop-activity-cross-workspace`; tagged client-side, not on the wire). Mirrors
 * `desktop_lib::dto::AggregatedOpLogEntryDto`. */
export interface AggregatedOpLogEntry {
	/** `"you@dev"` / `"agent:name@dev"` / `"external@dev"`. */
	principal: string;
	/** One-line human summary of the mutation. */
	op: string;
	/** Unix ms. */
	at_ms: number;
	/** `cli` / `tui` / `desktop` / `mcp` / `sync` / `external`; empty or absent for an old op. */
	source?: string;
	workspace_id: string;
	workspace_root: string;
}

/** Newest ops across every *registered* workspace (not just the one open in the main view),
 * merge-sorted newest-first and capped at 200 — same one-shot bounded-fetch contract as
 * `src/devices/api.ts::opLog()`'s single-workspace feed, not a live tail. A workspace whose root
 * no longer exists, or whose fetch fails, is silently skipped rather than failing the whole call. */
export function opLogAll(): Promise<AggregatedOpLogEntry[]> {
	return invoke("op_log_all");
}

/** One open (`!completed`) task line from some workspace's root `todo.txt`, tagged with enough
 * workspace identity to switch to it and label it. Mirrors
 * `desktop_lib::dto_universal::UniversalTaskDto`. */
export interface UniversalTask {
	workspace_id: string;
	workspace_root: string;
	line_number: number;
	priority: string | null;
	contexts: string[];
	description: string;
}

/** Every open task across every registered workspace's root `todo.txt` (ADR 0025, task
 * desktop-universal-view), unsorted — grouping by priority and filtering by `@context` are the
 * caller's job, over this flat list. Nested `ref:` sub-lists/notes aren't included (deliberate
 * scope cut, see todo.txt's `desktop-universal-view` entry); a workspace whose fetch fails is
 * silently absent from the result rather than failing the whole call. */
export function universalTasks(): Promise<UniversalTask[]> {
	return invoke("universal_tasks");
}

/** Tells the quick-add global hotkey's guard whether the main window's popover has an unsaved
 * edit right now (tasks/desktop-quick-add/notes.md: "if the main popover is open and dirty, the
 * hotkey focuses the main window instead"). Best-effort: callers swallow the rejection rather than
 * surfacing it, since a missed update here only affects which window the hotkey focuses next. */
export function setMainPopoverDirty(dirty: boolean): Promise<void> {
	return invoke("set_main_popover_dirty", { dirty });
}

File: apps/desktop/src-tauri/src/daemon.rs
//! The gRPC bridge to `txtodod`: a lazily-dialed, retrying client over the ADR 0010 unix socket,
//! plus the "make sure a daemon exists" bootstrap (design §5/§7, plan M7).
//!
//! The transport is unix-only for now (`#[cfg(unix)]`): Windows named pipes land with the
//! per-user service (`daemon-service-files`, plan M10). On every other target the functions here
//! return [`DaemonError::UnsupportedPlatform`] instead of assuming a `unix:` socket works.

mod spawn;
mod workspace;

pub use spawn::ensure_daemon;

use std::fmt;
use std::path::{Path, PathBuf};
use std::time::Duration;
use txtodo_proto::v1 as pb;

/// Per-attempt bound for `wait_until_ready`'s `Health` probe. Gated with the same `#[cfg(unix)]`
/// as its one use site (`connect`'s unix-socket dial): on Windows that arm is
/// `UnsupportedPlatform` and an ungated const is dead code, which CI's `-D warnings` rejects.
/// Ref: https://doc.rust-lang.org/reference/conditional-compilation.html#the-cfg-attribute
#[cfg(unix)]
const CONNECT_TIMEOUT: Duration = Duration::from_millis(500);
/// How many probe attempts `wait_until_ready` makes before giving up.
const MAX_CONNECT_RETRIES: u32 = 5;
/// Delay between probe attempts.
const RETRY_BACKOFF: Duration = Duration::from_millis(150);

/// Everything that can go wrong talking to `txtodod`. Callers turn this into a [`super::status::DaemonStatus`]
/// change, never a panic (design §5: "daemon-absent/disconnected is a state, never a panic").
#[derive(Debug)]
pub enum DaemonError {
    /// The endpoint or channel could not be built.
    Connect(tonic::transport::Error),
    /// The daemon answered with a gRPC error.
    Rpc(tonic::Status),
    /// `txtodod` could not be spawned.
    Spawn(std::io::Error),
    /// The client-side no-double-spawn lock could not be taken.
    Lock(std::io::Error),
    /// Waiting for the daemon to answer ran past its bound.
    Timeout,
    /// This platform has no working transport yet (Windows: M10).
    UnsupportedPlatform,
}

impl fmt::Display for DaemonError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            DaemonError::Connect(e) => write!(f, "connect: {e}"),
            DaemonError::Rpc(e) => write!(f, "daemon rpc: {e}"),
            DaemonError::Spawn(e) => write!(f, "spawn txtodod: {e}"),
            DaemonError::Lock(e) => write!(f, "spawn lock: {e}"),
            DaemonError::Timeout => write!(f, "daemon did not become ready in time"),
            DaemonError::UnsupportedPlatform => {
                write!(f, "Windows transport lands M10; unix sockets only for now")
            }
        }
    }
}

impl std::error::Error for DaemonError {}

/// First word of [`DaemonError::apply_text`] for a `FAILED_PRECONDITION`. The frontend's
/// `todotxt/saveBuffer.ts` matches it (`isStaleBase`); keep the two in step.
pub const FAILED_PRECONDITION_TOKEN: &str = "failed-precondition:";

impl DaemonError {
    /// What the frontend sees when `apply` is refused. A `FAILED_PRECONDITION` (the document moved
    /// under the caller: a stale `Replace` base, or a `TaskRef` whose line and id disagree) starts
    /// with a stable token, so the editor can fall back to a per-line save without matching
    /// tonic's prose. `tonic::Code`: https://docs.rs/tonic/latest/tonic/enum.Code.html
    pub fn apply_text(&self) -> String {
        match self {
            DaemonError::Rpc(s) if s.code() == tonic::Code::FailedPrecondition => {
                format!("{FAILED_PRECONDITION_TOKEN} {}", s.message())
            }
            other => other.to_string(),
        }
    }
}

impl From<tonic::Status> for DaemonError {
    fn from(status: tonic::Status) -> Self {
        DaemonError::Rpc(status)
    }
}

/// A client to `txtodod` over its ADR 0010 unix socket, dialed lazily — against the true global
/// daemon as of ADR 0025/M11 (task `desktop-workspace-switcher`), so one client now serves every
/// registered workspace by varying [`DaemonClient::selector`], not a fixed one dialed per socket.
///
/// `Clone` is cheap (a tonic `Channel` is a handle onto one shared connection,
/// https://docs.rs/tonic/latest/tonic/transport/struct.Channel.html#multiplexing-requests): every
/// command clones the client out of `AppState` and makes its RPC on the clone, so a slow call
/// never holds the lock the others need (`AppState::client_snapshot`).
#[derive(Clone)]
pub struct DaemonClient {
    inner: pb::txtodo_client::TxtodoClient<tonic::transport::Channel>,
    sock: PathBuf,
    /// Attached to every request below. `switch_workspace` is the only way to change it — every
    /// in-flight and future RPC on this client targets whichever workspace it last named.
    selector: Option<pb::WorkspaceSelector>,
}

impl DaemonClient {
    /// The socket path this client was built for.
    pub fn socket(&self) -> &Path {
        &self.sock
    }

    /// Points every future RPC at `root` instead of whatever workspace this client targeted
    /// before — no reconnect, since the global daemon is one process serving every registered
    /// workspace already. `root` auto-registers via the same `Path` selector bridge
    /// `WorkspaceCatalog::resolve` already proves at the daemon level.
    pub fn switch_workspace(&mut self, root: &Path) {
        self.selector = Some(pb::WorkspaceSelector {
            selector: Some(pb::workspace_selector::Selector::Path(
                root.display().to_string(),
            )),
        });
    }

    /// Builds a lazily-dialed channel to `sock`, targeting `selector` (`None` means "the sole
    /// open workspace", the same bridge every other client of the global daemon relies on until
    /// `switch_workspace` names one explicitly). This never blocks: the first RPC drives the
    /// actual unix-socket dial, bounded by [`CONNECT_TIMEOUT`].
    /// Ref: <https://docs.rs/tonic/latest/tonic/transport/struct.Endpoint.html#method.connect_lazy>
    #[cfg(unix)]
    pub async fn connect(
        sock: &Path,
        selector: Option<pb::WorkspaceSelector>,
    ) -> Result<DaemonClient, DaemonError> {
        let dst = format!("unix://{}", sock.display());
        let endpoint = tonic::transport::Endpoint::from_shared(dst)
            .map_err(DaemonError::Connect)?
            .connect_timeout(CONNECT_TIMEOUT);
        let channel = endpoint.connect_lazy();
        Ok(DaemonClient {
            inner: pb::txtodo_client::TxtodoClient::new(channel),
            sock: sock.to_path_buf(),
            selector,
        })
    }

    /// Stub for non-unix targets; plan M10 adds the Windows named-pipe transport.
    #[cfg(not(unix))]
    pub async fn connect(
        _sock: &Path,
        _selector: Option<pb::WorkspaceSelector>,
    ) -> Result<DaemonClient, DaemonError> {
        Err(DaemonError::UnsupportedPlatform)
    }

    /// Probes the daemon up to [`MAX_CONNECT_RETRIES`] times, [`RETRY_BACKOFF`] apart, so a
    /// freshly spawned daemon has time to bind the socket before the first real call. The probe is
    /// the registry-level `workspace_list`, never selector-scoped: it answers the moment the socket
    /// is bound, whereas a workspace-scoped `Health` would wait for that workspace's open (the
    /// daemon binds first and opens in the background, task `daemon-early-bind`) and an unselected
    /// one is refused while anything is still opening.
    pub async fn wait_until_ready(&mut self) -> Result<(), DaemonError> {
        let mut last: Option<DaemonError> = None;
        for attempt in 0..MAX_CONNECT_RETRIES {
            match self.workspace_list().await {
                Ok(_) => return Ok(()),
                Err(e) => last = Some(e),
            }
            if attempt + 1 < MAX_CONNECT_RETRIES {
                tokio::time::sleep(RETRY_BACKOFF).await;
            }
        }
        Err(last.unwrap_or(DaemonError::Timeout))
    }

    /// Every synced document with its current projection hash.
    pub async fn list_files(&mut self) -> Result<pb::ListFilesResponse, DaemonError> {
        let req = pb::ListFilesRequest {
            workspace: self.selector.clone(),
        };
        Ok(self.inner.list_files(req).await?.into_inner())
    }

    /// The exact bytes the daemon holds for one workspace-relative document path.
    pub async fn get_file(&mut self, path: &str) -> Result<pb::FileContents, DaemonError> {
        let req = pb::GetFileRequest {
            path: path.to_owned(),
            workspace: self.selector.clone(),
        };
        Ok(self.inner.get_file(req).await?.into_inner())
    }

    /// `get_file`, but against `selector` regardless of what this client's own `selector` field
    /// currently targets — never mutates it. The universal view's only caller: aggregating every
    /// registered workspace's `todo.txt` must not disturb whichever workspace the rest of the app
    /// (and this same connection) already has open.
    pub async fn get_file_for(
        &mut self,
        selector: pb::WorkspaceSelector,
        path: &str,
    ) -> Result<pb::FileContents, DaemonError> {
        let req = pb::GetFileRequest {
            path: path.to_owned(),
            workspace: Some(selector),
        };
        Ok(self.inner.get_file(req).await?.into_inner())
    }

    /// A change per reconcile or apply for `paths` (every document when empty).
    pub async fn watch(
        &mut self,
        paths: Vec<String>,
    ) -> Result<tonic::Streaming<pb::Change>, DaemonError> {
        let req = pb::WatchRequest {
            paths,
            workspace: self.selector.clone(),
        };
        Ok(self.inner.watch(req).await?.into_inner())
    }

    /// Intent-level mutations; the daemon turns them into ops.
    pub async fn apply(
        &mut self,
        mut req: pb::ApplyRequest,
    ) -> Result<pb::ApplyResponse, DaemonError> {
        req.workspace = self.selector.clone();
        Ok(self.inner.apply(req).await?.into_inner())
    }

    /// Ops newest first, filtered by path and/or task.
    pub async fn history(
        &mut self,
        mut req: pb::HistoryRequest,
    ) -> Result<pb::HistoryResponse, DaemonError> {
        req.workspace = self.selector.clone();
        Ok(self.inner.history(req).await?.into_inner())
    }

    /// Resolves one `needs_review` flag; maps to the daemon's `ResolveConflict` RPC.
    pub async fn resolve(
        &mut self,
        mut req: pb::ResolveRequest,
    ) -> Result<pb::ApplyResponse, DaemonError> {
        req.workspace = self.selector.clone();
        Ok(self.inner.resolve_conflict(req).await?.into_inner())
    }

    /// Open `needs_review` flags for `path` (plan M4): two devices rewrote the same word.
    pub async fn list_conflicts(
        &mut self,
        path: &str,
    ) -> Result<pb::ConflictsResponse, DaemonError> {
        let req = pb::ConflictsRequest {
            path: path.to_owned(),
            workspace: self.selector.clone(),
        };
        Ok(self.inner.list_conflicts(req).await?.into_inner())
    }

    /// `notes.md` for one task's `ref:` directory (plan M5); the daemon resolves the task to its
    /// directory, this client never touches the filesystem itself.
    pub async fn get_notes(&mut self, task: pb::TaskRef) -> Result<pb::NotesDoc, DaemonError> {
        let req = pb::GetNotesRequest {
            task: Some(task),
            workspace: self.selector.clone(),
        };
        Ok(self.inner.get_notes(req).await?.into_inner())
    }

    /// One whole-document edit to `notes.md`; the daemon derives the Loro text ops and lazily
    /// creates the `ref:` directory on the first edit (plan §3.2.4).
    pub async fn edit_notes(
        &mut self,
        mut req: pb::NotesEditRequest,
    ) -> Result<pb::ApplyResponse, DaemonError> {
        req.workspace = self.selector.clone();
        Ok(self.inner.edit_notes(req).await?.into_inner())
    }

    /// Starts a pairing handshake on this device and returns the QR payload (plan M4, design §4).
    pub async fn pair_offer(&mut self) -> Result<pb::PairOfferResponse, DaemonError> {
        let workspace = self.selector.clone();
        Ok(self
            .inner
            .pair_offer(pb::PairOfferRequest { workspace })
            .await?
            .into_inner())
    }

    /// Accepts a peer's scanned `PairOffer` (`code`) and begins the X25519 handshake; returns the
    /// 6-word SAS.
    pub async fn pair_accept(&mut self, code: String) -> Result<pb::PairResult, DaemonError> {
        let req = pb::PairAcceptRequest {
            code,
            workspace: self.selector.clone(),
        };
        Ok(self.inner.pair_accept(req).await?.into_inner())
    }

    /// Confirms the SAS shown to the human on this device; the group key lands only once both
    /// sides have confirmed. `own_device` is the human's answer to "is the other device your own?"
    /// (task default-workspace-pairing-consent): the default list merges only when both said yes.
    pub async fn pair_confirm_sas(
        &mut self,
        own_device: bool,
    ) -> Result<pb::PairResult, DaemonError> {
        let workspace = self.selector.clone();
        let req = pb::PairConfirmRequest {
            workspace,
            own_device,
        };
        Ok(self.inner.pair_confirm_sas(req).await?.into_inner())
    }

    /// Mints a new capability token from the design §6.2 scope/caveat grammar.
    pub async fn token_create(
        &mut self,
        mut req: pb::TokenCreateRequest,
    ) -> Result<pb::Token, DaemonError> {
        req.workspace = self.selector.clone();
        Ok(self.inner.token_create(req).await?.into_inner())
    }

    /// Tokens for this workspace, scopes included, secrets never returned.
    pub async fn token_list(&mut self) -> Result<pb::TokenListResponse, DaemonError> {
        let workspace = self.selector.clone();
        Ok(self
            .inner
            .token_list(pb::TokenListRequest { workspace })
            .await?
            .into_inner())
    }

    /// Revokes a token; the daemon refuses it on its next use.
    pub async fn token_revoke(
        &mut self,
        id: String,
    ) -> Result<pb::TokenRevokeResponse, DaemonError> {
        let req = pb::TokenRevokeRequest {
            id,
            workspace: self.selector.clone(),
        };
        Ok(self.inner.token_revoke(req).await?.into_inner())
    }

    /// Newest ops across every tracked file, at most 200, newest first: one bounded read, not a
    /// live tail (unlike [`DaemonClient::watch`]) — so this drains the whole stream before
    /// returning instead of forwarding it as an event stream.
    pub async fn op_log(&mut self) -> Result<Vec<pb::OpLogEntry>, DaemonError> {
        let workspace = self.selector.clone();
        self.op_log_with(workspace).await
    }

    /// As [`Self::op_log`], but targeting `workspace` explicitly rather than this client's own
    /// ambient `selector` — task `desktop-activity-cross-workspace`'s fan-out needs to read one
    /// workspace's op log while switched to (or connected with no selector on) another.
    pub async fn op_log_for(
        &mut self,
        workspace: pb::WorkspaceSelector,
    ) -> Result<Vec<pb::OpLogEntry>, DaemonError> {
        self.op_log_with(Some(workspace)).await
    }

    async fn op_log_with(
        &mut self,
        workspace: Option<pb::WorkspaceSelector>,
    ) -> Result<Vec<pb::OpLogEntry>, DaemonError> {
        let mut stream = self
            .inner
            .op_log_stream(pb::OpLogRequest { workspace })
            .await?
            .into_inner();
        let mut entries = Vec::new();
        while let Some(entry) = stream.message().await? {
            entries.push(entry);
        }
        Ok(entries)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a_failed_precondition_starts_with_the_stable_token() {
        let stale = DaemonError::Rpc(tonic::Status::failed_precondition("the document changed"));
        let text = stale.apply_text();
        assert!(text.starts_with(FAILED_PRECONDITION_TOKEN), "{text}");
        assert!(text.ends_with("the document changed"), "{text}");
    }

    #[test]
    fn any_other_refusal_keeps_its_plain_text() {
        let other = DaemonError::Rpc(tonic::Status::invalid_argument("bad line"));
        assert_eq!(other.apply_text(), other.to_string());
        assert!(!other.apply_text().contains(FAILED_PRECONDITION_TOKEN));
        assert_eq!(
            DaemonError::Timeout.apply_text(),
            DaemonError::Timeout.to_string()
        );
    }
}
